contracts/VotingRewards.sol

```diff
@@ -0,0 +1,502 @@
+        require(claimEnd > lastBlockDistributedReward, ERROR_EPOCH);
+
+        uint256 rewardAmount = _calculateReward(
+            _beneficiary,
+            startBlockNumberOfCurrentEpoch,
+            claimEnd,
+            missingVotesThreshold
+        );
+
+        uint64 currentBlockNumber = getBlockNumber64();
+        previousRewardsDistributionBlockNumber[_beneficiary] = currentBlockNumber;
+
+        addressRewards[_beneficiary].push(
+            Reward(
+                rewardAmount,
+                RewardState.Unlocked,
+                currentBlockNumber,
+                lockTime
+            )
+        );
+
+        baseVault.transfer(rewardsToken, rewardsVault, rewardAmount);
+        emit RewardDistributed(_beneficiary, rewardAmount, lockTime);
+    }
+
+    /**
+     * @notice collect rewards for an address if lockTime is passed since when tokens have been distributed
+     * @param _beneficiary address that should be fund with rewards
+     * @dev rewardsVault should have TRANSFER_ROLE permission
+     */
+    function collectRewardsFor(address _beneficiary) public {
+        uint64 timestamp = getBlockNumber64();
+        // prettier-ignore
+        Reward[] storage rewards = addressRewards[_beneficiary];
+
+        require(rewards.length > 0, ERROR_NO_REWARDS);
+
+        uint256 collectedRewards = 0;
+        for (uint256 i = 0; i < rewards.length; i++) {
+            if (
+                timestamp - rewards[i].lockBlock > rewards[i].lockTime &&
+                rewards[i].state != RewardState.Withdrawn
+            ) {
+                rewards[i].state = RewardState.Withdrawn;
+                rewardsVault.transfer(
+                    rewardsToken,
+                    _beneficiary,
+                    rewards[i].amount
+                );
+                collectedRewards = collectedRewards.add(1);
+
+                emit RewardCollected(_beneficiary, rewards[i].amount);
+            }
+        }
+
+        require(collectedRewards >= 1, ERROR_NO_REWARDS);
+    }
+
+    /**
+     * @notice Reward is calculated as the minimum balance between the
+     *         end of an epoch (now) and the balance at the first
+     *         vote in an epoch (in percentage) for each vote happened within the epoch
+     * @param _beneficiary beneficiary
+     * @param _fromBlock block from wich starting looking for votes
+     * @param _toBlock   block to wich stopping looking for votes
+     * @param _missingVotesThreshold number of vote to which is possible to don't vote
+     */
+    function _calculateReward(
+        address _beneficiary,
+        uint64 _fromBlock,
+        uint64 _toBlock,
+        uint256 _missingVotesThreshold
+    ) internal view returns (uint256) {
+        uint256 missingVotes = 0;
+        uint256 minimunBalance = 0;
+        bool forceFirstAssignment = true;
+
+        // voteId starts from 1 in DandelionVoting
+        for (
```

Comment:
Now that you've updated the loop in `_calculateReward`, why add 1 to `votes.length` then immediately subtract it again in the loop to get the vote ID?

Just use the following loop and get rid of `realVoteId` entirely:

```
for (uint256 voteId = dandelionVoting.votesLength(); voteId >= 1; voteId--) {
    ...
}
```

Reply:
yeah! correct!